autoload/lightline/bufferline.vim

```diff
@@ -94,21 +94,34 @@ function! s:select_buffers(before, current, after)
     return [a:before, a:current, a:after]
   endif
 
-  " Add one buffer on the right
+  let l:initial_right = 0
+  let l:right = 0
+  let l:left = 0
+
+  " Add one buffer on the right if there is enough space for it
   if len(l:after_lengths) > 0
-    let l:width -= l:after_lengths[0]
+    let [l:width, l:initial_right] = s:fit_lengths(l:after_lengths[:0], l:width)
   endif
 
-  " Add as many buffers as possible from on the left
-  let [l:width, l:left] = s:fit_lengths(l:before_lengths, l:width)
-  let l:before = a:before[-l:left:]
+  " Add as many buffers as possible on the left
+  " Don't forget to use the 'before' list in reversed order
+  let [l:width, l:left] = s:fit_lengths(reverse(l:before_lengths), l:width)
+  " Handle empty list carefully, slices are inclusive
+  let l:before = l:left == 0 ? [] : a:before[-l:left:]
+
+  " If one buffer on the right size was added, maybe more can fit?
+  if l:initial_right > 0
+    " Fill up the remaining space with buffers on the right
+    let [l:width, l:right] = s:fit_lengths(l:after_lengths[1:], l:width)
```

Comment:
I think we can reuse the `l:initial_right` variable here: `s:fit_lengths(l:after_lengths[l:initial_right:], l:width)`